### ISSUE TITLE
`Router::into_make_service_with_connect_info()` has unnecessary `Target` parameter that can fail to be inferred

### ISSUE THREAD

**Issue by lilyball** [tokio-rs/axum#859]
### Version

axum 0.4.8

### Crates

axum

### Description

[`Router::into_make_service_with_connect_info()`](https://docs.rs/axum/0.4.8/axum/struct.Router.html#method.into_make_service_with_connect_info)[^1] has a method signature

[^1]: Also [`Handler::into_make_service_with_connect_info()`](https://docs.rs/axum/0.4.8/axum/handler/trait.Handler.html#method.into_make_service_with_connect_info). I will only focus on `Router` here though.

```rust
pub fn into_make_service_with_connect_info<C, Target>(
    self
) -> IntoMakeServiceWithConnectInfo<Self, C>
where
    C: Connected<Target>
```

The `Target` parameter here only serves to ensure that `C` implements some `Connected<Target>`. From the sample in that docs, expected usage looks like

```rust
app.into_make_service_with_connect_info::<SocketAddr, _>()
```

The problem is, the second parameter can only be omitted if the `C` type only implements `Connected` for one single connection type. As soon as you have one `C` that implements `Connected` for multiple targets, that parameter needs to be specified explicitly. This is very annoying.

It's also unnecessary. [`IntoMakeServiceWithConnectInfo`](https://docs.rs/axum/0.4.8/axum/extract/connect_info/struct.IntoMakeServiceWithConnectInfo.html#) already only implements `Service<T> where C: Connected<T>`. Which means the bound on `into_make_service_with_connect_info()` isn't necessary for correctness. It probably does produce a better error message when the wrong type is used, but having to write out the connection type is frustrating, especially since I can write the wrong connection type on that method and it will still compile as long as my `C` type still implements `Connected` for the correct connection (i.e. I can write `app.into_make_service_with_connect_info::<SocketAddr, &'_ AddrStream>()` even though I'm using a different custom stream type for my actual server as long as `SocketAddr: Connected<&'_ MyCustomStream>`).

I'm not really sure why this bound exists here. axum generally defers bounds until the point at which they're absolutely necessary, it's usually possible to construct a type with the wrong parameters only to have it fail to implement the right trait as a result, so it's surprising to me that `into_make_service_with_connect_info()` bothers with this bound.

#### Proposed solution

Remove the `C: Connected<Target>` bound on `into_make_service_with_connect_info()`. The resulting type will still enforce this bound on its `Service` impl. The type parameter has to go as well. Unfortunately it looks like Rust doesn't allow defaulted type parameters on methods (?!?) so we can't just leave it unused with a default, which does mean this is backwards-incompatible.

#### Alternative solution

Add the `Target` parameter as an actual parameter of `IntoMakeServiceWithConnectInfo` and only implement `Service` in terms of that instead of making it generic. This way the call to `Server::serve(app)` will infer that type parameter if it's otherwise ambiguous. The downside is now the same `IntoMakeServiceWithConnectInfo` can no longer be used with different connection types. It's also backwards-incompatible so I don't think there's any benefit to this approach.

**Comment by davidpdrsn** [comment]
> As soon as you have one C that implements Connected for multiple targets, that parameter needs to be specified explicitly. This is very annoying.

What is your use case for this?

> It probably does produce a better error message when the wrong type is used, but having to write out the connection type is frustrating

and

> I'm not really sure why this bound exists here. axum generally defers bounds until the point at which they're absolutely necessary, it's usually possible to construct a type with the wrong parameters only to have it fail to implement the right trait as a result, so it's surprising to me that into_make_service_with_connect_info() bothers with this bound.

Giving a better error message is exactly the reason the bound is there. I don't generally prefer defering bounds to the trait impls because it hurts error messages. Very often axum's users are beginners so we try and cater to those by giving good errors where possible, which it isn't always however.

Also, please don't write stuff like "I'm not really sure why this bound exists here". It comes off as rude. I know thats probably not your intention, but still. Instead you I'd appreciate it if you wrote "Whats the reason for the bound? To improve errors messages?".

With all that said I'm fine with removing the bound. I'm mostly curious about what your use case is.

**Comment by lilyball** [comment]
Sorry for not replying. My use-case was I have a custom connect info type that implements `Connected` for two separate stream types, which meant I had to specify one of those two stream types on every call to `into_make_service_with_connect_info` even though doing so did not actually enforce that this stream type was used with the resulting `MakeService`.

I see that you already fixed this in v0.5.0, thank you.
